### ISSUE TITLE
WeakRef::RefError: Invalid Reference - probably recycled

### ISSUE THREAD

**Issue by GregOnNet** [nanoc/nanoc#863]
Hi,

first thanks for the great and consistent work on nanoc.
Me and my team really appreciate it.

**Problem**

I create the used items inside a `preprocess` rule.
After upgrading to Nanoc 4.2.0 I experience the following error when I access `@item` inside a template:

````
Message:

WeakRef::RefError: Invalid Reference - probably recycled

Compilation stack:

  (empty)
````

**Environment**

```
Nanoc 4.2.0 © 2007-2016 Denis Defreyne.
Running ruby 2.2.4 (2015-12-16) on x64-mingw32 with RubyGems 2.4.5.1.
```

**crash.log**
[crash.txt](https://github.com/nanoc/nanoc/files/305231/crash.txt)

Let me know if I can provide further information  to this topic

Kind Regards
Gregor

**Comment by agross** [comment]
Here's another stack trace with less haml and more nanoc:

[crash.log.txt](https://github.com/nanoc/nanoc/files/305268/crash.log.txt)

**Comment by ddfreyne** [comment]
Yup, I am getting this one as well. This is caused by the recent change of #846, though I am not quite sure why.

**Comment by ddfreyne** [comment]
Potential fix in #865.

**Comment by agross** [comment]
The only reason for this to fail seems be that GC occurs between the check for `weakref_alive?` and the actual retrieval of the value.

**Comment by agross** [comment]
Probably it's best to [disable GC during check and retrieval](https://books.google.de/books?id=H0SNJTGe-RIC&pg=PA571&lpg=PA571&dq=ruby+weakref+weakref_alive+gc+between&source=bl&ots=9gVA7m0UUv&sig=QSmgj8Fkz1rKm5RDlOu19TycRyc&hl=en&sa=X&ved=0ahUKEwj2tPyd-ZjNAhUHvBQKHdZcC_MQ6AEIJDAB#v=onepage&q&f=false)?

**Comment by agross** [comment]
Yes, it seems GC is the culprit.

I added this:

```diff
+       require 'pry-byebug'
+       prestat = GC.stat
        if method_cache.key?(args) && method_cache[args].weakref_alive?
+         begin
            method_cache[args].value
+          rescue => e
+            errstat = GC.stat
+            binding.pry
+            raise
+          end
        else
          send(original_method_name, *args).tap do |r|
            method_cache[args] = WeakRef.new(Wrapper.new(r))
          end
        end
```

When the debugger breaks I printed the GC count:
```
[3] pry(#<Nanoc::Int::OutdatednessChecker>)> prestat[:count]
50
[3] pry(#<Nanoc::Int::OutdatednessChecker>)> errstat[:count]
51
```

**Comment by ddfreyne** [comment]
Correct. I have a fix in #865, which uses `rescue WeakRef::RefError` rather than checking `#weakref_alive?`.
